File: app/src/main/java/com/usharik/app/UiLanguageManager.java
package com.usharik.app;

import android.content.Context;
import android.content.SharedPreferences;

import androidx.appcompat.app.AppCompatDelegate;
import androidx.core.os.LocaleListCompat;

import com.usharik.app.fragment.SettingsFragment;

public final class UiLanguageManager {
    public static final String UI_LANGUAGE_KEY = "uiLanguage";

    private UiLanguageManager() {
    }

    public static boolean hasSavedLanguage(Context context) {
        return getPreferences(context).contains(UI_LANGUAGE_KEY);
    }

    public static UiLanguage getSavedLanguage(Context context) {
        return UiLanguage.fromPreference(getPreferences(context).getString(UI_LANGUAGE_KEY, null));
    }

    public static UiLanguage getSelectedLanguage(Context context) {
        return hasSavedLanguage(context) ? getSavedLanguage(context) : UiLanguage.SYSTEM;
    }

    public static String getSelectedLanguageLabel(Context context) {
        return getSelectedLanguage(context).displayName(context);
    }

    public static UiLanguage[] getAvailableLanguages() {
        return UiLanguage.values();
    }

    public static CharSequence[] getLanguageLabels(Context context) {
        UiLanguage[] options = getAvailableLanguages();
        CharSequence[] labels = new CharSequence[options.length];
        for (int i = 0; i < options.length; i++) {
            labels[i] = options[i].displayName(context);
        }
        return labels;
    }

    public static int indexOf(UiLanguage target) {
        UiLanguage[] options = getAvailableLanguages();
        for (int i = 0; i < options.length; i++) {
            if (options[i] == target) {
                return i;
            }
        }
        return 0;
    }

    public static boolean applySavedLanguage(Context context) {
        if (!hasSavedLanguage(context)) {
            return false;
        }
        return applyLanguage(getSavedLanguage(context));
    }

    public static boolean saveAndApplyLanguage(Context context, UiLanguage uiLanguage) {
        getPreferences(context).edit().putString(UI_LANGUAGE_KEY, uiLanguage.preferenceValue()).apply();
        return applyLanguage(uiLanguage);
    }

    public static boolean applyLanguage(UiLanguage uiLanguage) {
        LocaleListCompat newLocales = toLocaleList(uiLanguage);
        LocaleListCompat currentLocales = AppCompatDelegate.getApplicationLocales();
        if (currentLocales.equals(newLocales)) {
            return false;
        }
        AppCompatDelegate.setApplicationLocales(newLocales);
        return true;
    }

    private static LocaleListCompat toLocaleList(UiLanguage uiLanguage) {
        return uiLanguage == UiLanguage.SYSTEM
                ? LocaleListCompat.getEmptyLocaleList()
                : LocaleListCompat.forLanguageTags(uiLanguage.languageTags());
    }

    private static SharedPreferences getPreferences(Context context) {
        return context.getSharedPreferences(SettingsFragment.SHARED_PREFERENCES, Context.MODE_PRIVATE);
    }
}

File: app/src/main/kotlin/com/usharik/app/UiLanguageManager.kt
package com.usharik.app

import android.content.Context
import androidx.appcompat.app.AppCompatDelegate
import androidx.core.os.LocaleListCompat

object UiLanguageManager {
    const val UI_LANGUAGE_KEY = "uiLanguage"
    fun hasSavedLanguage(context: Context) = preferences(context).contains(UI_LANGUAGE_KEY)
    fun getSavedLanguage(context: Context) = UiLanguage.fromPreference(preferences(context).getString(UI_LANGUAGE_KEY, null))
    fun getSelectedLanguage(context: Context) = if (hasSavedLanguage(context)) getSavedLanguage(context) else UiLanguage.SYSTEM
    fun getSelectedLanguageLabel(context: Context) = getSelectedLanguage(context).displayName(context)
    fun getAvailableLanguages() = UiLanguage.entries.toTypedArray()
    fun getLanguageLabels(context: Context): Array<CharSequence> = getAvailableLanguages().map { it.displayName(context) }.toTypedArray()
    fun indexOf(target: UiLanguage) = getAvailableLanguages().indexOf(target).coerceAtLeast(0)
    fun applySavedLanguage(context: Context) = hasSavedLanguage(context) && applyLanguage(getSavedLanguage(context))
    fun saveAndApplyLanguage(context: Context, language: UiLanguage): Boolean { preferences(context).edit().putString(UI_LANGUAGE_KEY, language.preferenceValue()).apply(); return applyLanguage(language) }
    fun applyLanguage(language: UiLanguage): Boolean {
        val locales = if (language == UiLanguage.SYSTEM) LocaleListCompat.getEmptyLocaleList() else LocaleListCompat.forLanguageTags(language.languageTags())
        if (AppCompatDelegate.getApplicationLocales() == locales) return false
        AppCompatDelegate.setApplicationLocales(locales); return true
    }
    private fun preferences(context: Context) = context.getSharedPreferences(App.PREFS_NAME, Context.MODE_PRIVATE)
}
